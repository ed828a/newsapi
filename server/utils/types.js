const Categories = Object.freeze({
  Business: "business",
  Entertainment: "entertainment",
  General: "general",
  Health: "health",
  Science: "science",
  Sports: "sports",
  Technology: "technology",
  Topheadline: "topheadline"
});

export { Categories };
