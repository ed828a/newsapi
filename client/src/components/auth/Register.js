import React, { Component, Fragment } from "react";
import axios from "axios";
import config from "../../config";
import { Link } from "react-router-dom";

const url = `${config.BASE_URL}/api/auth/register`;

export default class Register extends Component {
  state = {
    username: "",
    password: "",
    error: "",
    validLogin: false
  };

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };
  handleSubmit = event => {
    axios
      .post(url, {
        username: this.state.username,
        password: this.state.password
      })
      .then(data => {
        console.log(data);
        this.setState({
          validLogin: true,
          username: "",
          password: "",
          error: ""
        });
      })
      .catch(err => {
        this.setState(error => {
          const errorkey = Object.keys(err.response.data.errors);
          const errorval = Object.values(err.response.data.errors);
          const msg = errorkey[0] + " " + errorval[0];
          return { error: msg };
        });
      });

    event.preventDefault();
  };

  render() {
    return (
      <div id="main">
        <div className="container">
          <div className="row">
            <div className="marginleft33">
              <h1>Register</h1>
            </div>
          </div>

          {this.state.error ? (
            <Fragment>
              <div className="row">&nbsp;</div>
              <div className="row marginleft40" style={{ color: "red" }}>
                {this.state.error}
              </div>
            </Fragment>
          ) : (
            ""
          )}
          {this.state.validLogin ? (
            <Fragment>
              <div className="row">&nbsp;</div>

              <div className="marginleft40" style={{ color: "green" }}>
                User successFully Registered.Please log in &nbsp;
                <Link to="/login" style={{ textDecoration: "underline" }}>
                  <b>here</b>
                </Link>
                <div className="row">&nbsp;</div>
              </div>
            </Fragment>
          ) : (
            ""
          )}
          <div className="row">
            <div className="marginleft33">Username :</div>
            <div className="cols-md-4">
              <input
                type="text"
                name="username"
                value={this.state.username}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="marginleft33">Password : </div>
            <div className="cols-md-4">
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.handleChange}
              />
            </div>
          </div>

          <div className="row">
            <div className="marginleft40" />
            <div className="cols-md-4">
              <input
                className="btnlogin"
                type="button"
                name="button"
                value="Register"
                onClick={this.handleSubmit}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}
