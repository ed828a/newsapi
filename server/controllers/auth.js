import User from "../models/user";
import passport from "passport";
import {
  userToJSON,
  getUserToken,
  encryptPassword,
  isPasswordAllowed
} from "../utils/auth";

const authUserJson = user => ({
  ...userToJSON(user),
  token: getUserToken(user)
});
async function register(req, res) {
  const { username, password } = req.body;
  if (!username) {
    return res.status(422).json({ errors: { username: `can't be blank` } });
  }
  if (!password) {
    return res.status(422).json({ errors: { password: `can't be blank` } });
  }

  if (!isPasswordAllowed(password)) {
    return res.status(422).json({
      errors: {
        password: ` must be greater than 6 character.Must contain special
      character and letter`
      }
    });
  }

  const existingUser = await User.findOne({ username });

  if (existingUser) {
    return res.status(422).json({ errors: { username: "taken" } });
  }
  const encPassword = await encryptPassword(password);

  const newUser = await User.create({
    username,
    password: encPassword
  });

  return res.json({ user: authUserJson(newUser) });
}

async function login(req, res, next) {
  const { username, password } = req.body;
  if (!username) {
    return res.status(422).json({ errors: { username: `can't be blank` } });
  }
  if (!password) {
    return res.status(422).json({ errors: { password: `can't be blank` } });
  }

  const { user, info } = await authenticate(req, res, next);
  if (user) {
    return res.json({ user: authUserJson(user) });
  } else {
    return res.status(422).json(info);
  }
}
function authenticate(req, res, next) {
  return new Promise((resolve, reject) => {
    passport.authenticate("local", { session: false }, (err, user, info) => {
      if (err) {
        reject(err);
      } else {
        resolve({ user, info });
      }
    })(req, res, next);
  });
}
async function me(req, res) {}

export { me, login, register };
