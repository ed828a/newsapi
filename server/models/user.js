import mongoose from "mongoose";
const { Schema } = mongoose;
const userSchema = new Schema({
  username: {
    type: String,
    required: [true, "user must have source"]
  },
  password: {
    type: String,
    required: [true, "user must have password"]
  },
  createdAt: {
    type: Date,
    default: Date.now,
    index: true
  },
  isActive: {
    type: Boolean,
    default: true
  }
});

export default mongoose.model("User", userSchema);
