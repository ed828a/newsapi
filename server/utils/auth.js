import jwt from "jsonwebtoken";
import bcrypt from "bcryptjs";
import config from "../config";
import expressJWT from "express-jwt";
import LocalStrategy from "passport-local";
import User from "../models/user";
import { omit } from "lodash";
const twoHours = 60 * 60 * 24;
const requireTime = Date.now();

const now = () =>
  process.env.NODE_ENV === "production" ? Date.now() : requireTime;
const issuedAt = Math.floor(now() / 1000);

const authMiddleware = {
  required: expressJWT({
    secret: config.secret
  }),
  optional: expressJWT({
    secret: config.secret,
    credentialsRequired: false
  })
};

async function encryptPassword(plainText) {
  const salt = await bcrypt.genSalt(10);
  const encPassword = await bcrypt.hash(plainText, salt);
  return encPassword;
}

async function comparePassword(plainPassword, encryptedPassword) {
  try {
    const result = await bcrypt.compare(plainPassword, encryptedPassword);
    return result;
  } catch (error) {
    return false;
  }
}
function isPasswordAllowed(password) {
  return password.length > 6 && /\d/.test(password) && /\D/.test(password);
}

function getLocalStrategy() {
  return new LocalStrategy(async (username, password, done) => {
    let user;

    try {
      user = await User.findOne({ username }).lean();
    } catch (error) {
      return done(error);
    }

    if (!user || (await comparePassword(password, user.password)) === false) {
      return done(null, false, {
        errors: { "username or password": "is invalid" }
      });
    }

    return done(null, userToJSON(user));
  });
}

function getUserToken({ _id, username }) {
  const issuedAt = Math.floor(now());
  return jwt.sign(
    {
      id: _id,
      username,
      iat: issuedAt,
      exp: issuedAt + twoHours
    },
    config.secret
  );
}
function userToJSON(user) {
  return omit(user, ["createdAt", "password", "__v", "isActive"]);
}
export {
  authMiddleware,
  getUserToken,
  getLocalStrategy,
  isPasswordAllowed,
  encryptPassword,
  comparePassword,
  userToJSON
};
