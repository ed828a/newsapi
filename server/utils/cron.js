import cron from "node-cron";
import Article from "../models/article";
import { Categories } from "./types";
import NewsAPI from "newsapi";
import config from "../config";

const newsapi = new NewsAPI(config.API_KEY);

function insertArticles(articles, category = Categories.Topheadline) {
  let promises = articles.map(article => {
    let doc = {
      category,
      source: article.source.name,
      author: article.author,
      description: article.description,
      url: article.url,
      urltoimage: article.urlToImage,
      publishedAt: article.publishedAt,
      content: article.content
    };
    return Article.create(doc)
      .then(data => {
        console.log("successfully updated");
      })
      .catch(err => {
        console.log(err.message);
      });
  });
  return promises;
}
function categoryNewsCron(category) {
  //return new Promise((resolve, reject) => {
  cron.schedule("*/20 * * * *", () => {
    console.log("running every minute 15 for category" + category);
    Article.remove({ category })
      .then(() => {
        newsapi.v2
          .topHeadlines({
            category,
            language: "en",
            country: "us"
          })
          .then(response => {
            if (response.articles && response.articles.length > 0) {
              let promises = insertArticles(response.articles, category);
              Promise.all(promises)
                .then(() => console.log("cron completed successfully "))
                .catch(err => {
                  console.log("failed", err);
                });
              //resolve(promises);
            }
          });
      })
      .catch(err => {
        console.log(err);
        // reject(err);
      });
  });
  // });
}

function topNewsDataCron() {
  cron.schedule("*/10 * * * *", () => {
    console.log("running a task every 5 minute");
    Article.remove({ category: Categories.Topheadline })
      .then(() => {
        newsapi.v2
          .topHeadlines({
            language: "en",
            country: "us"
          })
          .then(response => {
            console.log(response.articles);
            if (response.articles && response.articles.length > 0) {
              let promises = insertArticles(response.articles);
              Promise.all(promises).then(() =>
                console.log("cron completed successfully")
              );
            }
          });
      })
      .catch(err => console.log(err));
  });
}

function setUpCronJob() {
  topNewsDataCron();
  const categories = Object.values(Categories);
  categories.map(category => {
    categoryNewsCron(category);
  });
}

export default setUpCronJob;
