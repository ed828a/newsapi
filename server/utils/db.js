import mongoose from "mongoose";
import settings from "../config";
mongoose.Promise = global.Promise;

export const connect = () =>
  mongoose
    .connect(
      settings.mongoURI,
      { useNewUrlParser: true }
    )
    .then(() => {
      console.log("connected to mongo");
    })
    .catch(err => {
      console.log("connection error" + err);
    });
