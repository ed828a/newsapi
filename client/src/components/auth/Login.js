import React, { Component, Fragment } from "react";
import { authService } from "../../utils/authservice";

export default class Login extends Component {
  state = {
    username: "",
    password: "",
    error: ""
  };

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };
  handleSubmit = event => {
    authService
      .login(this.state.username, this.state.password)
      .then(data => this.props.history.push("/"))
      .catch(err => {
        this.setState(error => {
          const errorkey = Object.keys(err.response.data.errors);
          const errorval = Object.values(err.response.data.errors);
          const msg = errorkey[0] + " " + errorval[0];
          return { error: msg };
        });
      });

    event.preventDefault();
  };

  render() {
    return (
      <div id="main">
        <div className="container">
          <div className="row">
            <div className="marginleft33">
              <h1>Login</h1>
            </div>
          </div>

          {this.state.error ? (
            <Fragment>
              <div className="row">&nbsp;</div>
              <div className="row marginleft40" style={{ color: "red" }}>
                {this.state.error}
              </div>
            </Fragment>
          ) : (
            ""
          )}
          <div className="row">
            <div className="marginleft33">Username :</div>
            <div className="cols-md-4">
              <input
                type="text"
                name="username"
                value={this.state.username}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="marginleft33">Password : </div>
            <div className="cols-md-4">
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="marginleft40" />
            <div className="cols-md-4">
              <input
                className="btnlogin"
                type="button"
                name="button"
                value="Login"
                onClick={this.handleSubmit}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}
