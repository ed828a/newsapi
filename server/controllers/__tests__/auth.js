import faker from "faker";
import passport from "passport";
import User from "../../models/user";
import * as authController from "../auth";
import user from "../../models/user";
import { getUserToken } from "../../utils/auth";
function setup() {
  const req = {
    body: {}
  };
  const res = {
    status: jest.fn(function status() {
      return this;
    }),
    json: jest.fn(function json() {
      return this;
    }),
    send: jest.fn(function send() {
      return this;
    })
  };

  const next = jest.fn();
  return { req, res, next };
}

describe("Register", () => {
  var userObj = {
    _id: faker.random.uuid(),
    username: faker.internet.userName()
  };
  it("should require username", async () => {
    const { req, res, next } = setup();
    req.body.password = faker.internet.password();
    const result = await authController.register(req, res, next);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(422);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      errors: { username: expect.any(String) }
    });
    expect(res.json.mock.calls[0][0]).toMatchSnapshot();
  });

  it("should require password", async () => {
    const { req, res, next } = setup();
    req.body = { username: faker.internet.userName() };
    await authController.register(req, res, next);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(422);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      errors: { password: expect.any(String) }
    });
    expect(res.json.mock.calls[0][0]).toMatchSnapshot();
  });
  it("should require password with specific length", async () => {
    const { req, res, next } = setup();
    req.body = { username: faker.internet.userName(), password: "123" };
    await authController.register(req, res, next);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(422);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      errors: { password: expect.any(String) }
    });
    expect(res.json.mock.calls[0][0]).toMatchSnapshot();
  });
  it("should register and return user", async () => {
    const { req, res, next } = setup();
    req.body = {
      username: faker.internet.userName(),
      password: faker.internet.password()
    };
    jest.spyOn(User, "findOne");
    jest.spyOn(User, "create");
    User.findOne.mockImplementationOnce(() => {
      return Promise.resolve(0);
    });
    User.create.mockImplementationOnce(() => {
      const response = {
        ...userObj,
        password: faker.internet.password(),
        isActive: true,
        createdAt: "2018-11-28 22:10:54.737",
        __v: 0
      };
      return Promise.resolve(response);
    });
    const result = await authController.register(req, res, next);
    expect(res.json).toHaveBeenCalledTimes(1);
    const expObject = {
      ...userObj,
      token: expect.any(String)
    };

    expect(result.json.mock.calls[0][0].user).toMatchObject(expObject);

    User.findOne.mockRestore();
    User.create.mockRestore();
  });

  it("should require unique username", async () => {
    const { req, res, next } = setup();
    req.body = {
      username: userObj.username,
      password: faker.internet.password()
    };
    jest.spyOn(User, "findOne");
    User.findOne.mockImplementationOnce(() => {
      return Promise.resolve(userObj);
    });
    await authController.register(req, res, next);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(422);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      errors: { username: expect.any(String) }
    });
    expect(res.json.mock.calls[0][0]).toMatchSnapshot();
    User.findOne.mockRestore();
  });
});

describe("login", () => {
  it("should require username to login", async () => {
    const { req, res, next } = setup();
    req.body = { password: faker.internet.password() };
    await authController.login(req, res, next);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(422);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      errors: { username: expect.any(String) }
    });
    expect(res.json.mock.calls[0][0]).toMatchSnapshot();
  });
  it("should require password to login", async () => {
    const { req, res, next } = setup();
    req.body = { username: faker.internet.userName() };
    await authController.login(req, res, next);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(422);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      errors: { password: expect.any(String) }
    });
    expect(res.json.mock.calls[0][0]).toMatchSnapshot();
  });
  it("should throw passport authenticate error", async () => {
    jest.spyOn(passport, "authenticate");
    const fakeError = new Error("fake error");
    passport.authenticate.mockImplementationOnce((type, options, cb) => {
      cb(fakeError);
    });
    const { req, res, next } = setup();
    req.body = {
      username: faker.internet.userName(),
      password: faker.internet.password()
    };
    const error = await authController.login(req, res, next).catch(e => e);
    expect(error).toBe(fakeError);
    passport.authenticate.mockRestore();
  });

  it("should login and register with valid credentials", async () => {
    jest.spyOn(passport, "authenticate");

    const loginData = {
      username: faker.internet.userName(),
      password: faker.internet.password()
    };
    const fakeUser = {
      _id: faker.random.uuid(),
      ...loginData
    };

    const fakeToken = getUserToken(fakeUser);

    passport.authenticate.mockImplementationOnce((type, options, cb) => {
      cb(null, fakeUser);
    });
    const { req, res, next } = setup();
    req.body = loginData;
    await authController.login(req, res, next);
    expect(res.status).not.toHaveBeenCalled();
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toHaveBeenCalledWith({
      user: { _id: fakeUser._id, token: fakeToken, username: fakeUser.username }
    });
    passport.authenticate.mockRestore();
  });
});
