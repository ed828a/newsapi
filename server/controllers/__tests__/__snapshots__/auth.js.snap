// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should require password 1`] = `
Object {
  "errors": Object {
    "password": "can't be blank",
  },
}
`;

exports[`Register should require password with specific length 1`] = `
Object {
  "errors": Object {
    "password": "password doesn't meeet requirements",
  },
}
`;

exports[`Register should require unique username 1`] = `
Object {
  "errors": Object {
    "username": "taken",
  },
}
`;

exports[`Register should require username 1`] = `
Object {
  "errors": Object {
    "username": "can't be blank",
  },
}
`;

exports[`login should require password to login 1`] = `
Object {
  "errors": Object {
    "password": "can't be blank",
  },
}
`;

exports[`login should require username to login 1`] = `
Object {
  "errors": Object {
    "username": "can't be blank",
  },
}
`;
